libecole/tests/src/reward/test-nnodes.cpp

```diff
@@ -0,0 +1,85 @@
+#include <memory>
+#include <string>
+#include <tuple>
+
+#include <catch2/catch.hpp>
+
+#include "ecole/environment/branching.hpp"
+#include "ecole/observation/nothing.hpp"
+#include "ecole/reward/nnodes.hpp"
+
+#include "conftest.hpp"
+#include "reward/unit-tests.hpp"
+
+using namespace ecole;
+
+TEST_CASE("NNodes unit tests", "[unit][reward]") {
+	reward::unit_tests(reward::NNodes{});
+}
+
+TEST_CASE(
+	"NNodes returns the difference in the total number of processed nodes between two states",
+	"[reward]") {
+
+	auto reward_func = reward::NNodes{};
+	auto model = get_solving_model();
+	reward_func.reset(model);
+
+	SECTION("NNodes is positive") { REQUIRE(reward_func.obtain_reward(model) >= 0); }
+
+	SECTION("NNodes is zero if nothing happended between two states") {
+		reward_func.obtain_reward(model);
+		REQUIRE(reward_func.obtain_reward(model) == 0);
+	}
+
+	SECTION("Reset NNodes counter") {
+		reward_func.reset(model);
+		REQUIRE(reward_func.obtain_reward(model) >= 0);
+	}
+
+	SECTION("NNodes is consistent when used in a Branching environment") {
```

Comment:
I would prefer that reward functions be tested without using environments, as it brings extra entanglements between the classes which slow down the evolution of the code base (In the past I would have to go though many fixes in the tests for a single change to the environment).

Another way to put it is: it is not the responsibility of the reward_function that other classes use it correctly.
In practice, it could be useful to have integration tests that use different environments, with different fuctions etc., but I would prefer to do isolate it, has done with unit tests.

Reply:
Then the tests here can only be very basic ones. This is indeed a high-level test, which uses the logic of both the branching environment and the NNodes reward to make sure everything is consistent. Where should I put such a test then ?

Reply:
I don't have solution for now. Later we'll put some structure to have integration tests without too much copy paste.

In the meantime, either leave it out (other reward functions don't have it) or make it a separate `TEST_CASE` so it easier to refactor in the future.

But note that when you defined `class NNodes`, you inherited from `RewardFunction` and marked method `override` so simply compiling is already testing that it will be compatible with environments.